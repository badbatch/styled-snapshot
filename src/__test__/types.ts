import { Func } from "@repodog/types";
import { ComponentType } from "enzyme";
import { ReactElement, ReactFragment, ReactPortal, Ref } from "react";
import { FlattenSimpleInterpolation } from "styled-components";
import { ContextConsumerElement, ContextProviderElement, ForwardRefElement, MemoElement } from "../types";

export interface ComponentProps {
  callback?: Func;
  Component?: ComponentType<{}>;
  consumer?: ContextConsumerElement;
  css?: FlattenSimpleInterpolation;
  decorator?: ForwardRefElement;
  element?: ReactElement;
  forwardRef?: ForwardRefElement;
  fragment?: ReactFragment;
  innerRef?: Ref<HTMLDivElement>;
  memo?: MemoElement;
  portal?: ReactPortal;
  provider?: ContextProviderElement;
  renderProp?: (...args: any[]) => ReactElement;
  styled?: ComponentType<{}>;
  styledFunctionComponent?: ReactElement;
  styledStyledComponent?: ReactElement;
  styles?: FlattenSimpleInterpolation;
  value?: string;
}
