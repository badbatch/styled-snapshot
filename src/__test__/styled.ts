import styled, { css } from "styled-components";

export const StyledDiv = styled.div`
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  box-sizing: border-box;
  position: relative;
`;

export const StyledList = styled.ul`
  margin: 0;
  padding: 0;
`;

const itemStyles = css`
  ${props => (props.theme ? "position: relative" : "position: absolute")};
  padding: ${props => props.theme && "12px"};
`;

export const StyledItem = styled.li`
  margin-left: 12px;
  ${() => itemStyles};

  &:last-child {
    margin-left: 0;
  }
`;
