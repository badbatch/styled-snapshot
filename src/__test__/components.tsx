import { ObjectMap } from "@repodog/types";
import PropTypes from "prop-types";
import React, { Component, ComponentType, forwardRef, memo } from "react";
import { createPortal } from "react-dom";
import styled, { ThemeContext, ThemeProvider, withTheme } from "styled-components";
import { StyledDiv, StyledItem, StyledList } from "./styled";
import { ComponentProps } from "./types";

export class ClassComponent extends Component<ComponentProps> {
  public static displayName = "ClassComponent";

  public static propTypes = {
    Component: PropTypes.elementType,
  };

  // tslint:disable-next-line prefer-function-over-method
  public render() {
    return this.props.children;
  }
}

export const ContextComponent = () => (
  <ThemeContext.Provider value="dark">
    <ContextConsumerComponent />
  </ThemeContext.Provider>
);

ContextComponent.displayName = "ContextComponent";

export const ContextConsumerComponent = () => (
  <ThemeContext.Consumer>
    {value => (
      <ClassComponent value={value}>
        <FunctionComponent />
      </ClassComponent>
    )}
  </ThemeContext.Consumer>
);

ContextConsumerComponent.displayName = "ContextConsumerComponent";

export const ContextProviderComponent = () => (
  <ThemeContext.Provider value="dark">
    <ClassComponent>
      <FunctionComponent />
    </ClassComponent>
  </ThemeContext.Provider>
);

ContextProviderComponent.displayName = "ContextProviderComponent";

export const ForwardRefComponent = forwardRef<HTMLDivElement, ComponentProps>((props, ref) => (
  <ClassComponent {...props} innerRef={ref}>
    <FunctionComponent />
  </ClassComponent>
));

ForwardRefComponent.displayName = "ForwardRefComponent";

export const FunctionComponent = (props: ComponentProps) => <StyledDiv {...props} />;
FunctionComponent.displayName = "FunctionComponent";

FunctionComponent.propTypes = {
  Component: PropTypes.elementType,
};

export const MemoComponent = memo(FunctionComponent);
MemoComponent.displayName = "MemoComponent";

export const portal = createPortal(
  <ClassComponent>
    <FunctionComponent />
  </ClassComponent>,
  document.createElement("div"),
);

export const PortalComponent = () => portal;
PortalComponent.displayName = "PortalComponent";

export const SCComponent = () => (
  <StyledDiv>
    <ClassComponent>
      <FunctionComponent />
    </ClassComponent>
    <FunctionComponent />
    <StyledList>
      <StyledItem>One</StyledItem>
      <StyledItem>Two</StyledItem>
      <StyledItem>Three</StyledItem>
      <StyledItem>Four</StyledItem>
    </StyledList>
  </StyledDiv>
);

SCComponent.displayName = "SCComponent";

export const SCContextComponent = () => (
  <ThemeProvider theme={{ type: "light" }}>
    <WithThemeClassComponent />
  </ThemeProvider>
);

SCContextComponent.displayName = "SCContextComponent";

export const WithThemeClassComponent = withTheme<ComponentType<any>>(ClassComponent); // tslint:disable-line no-any

export const SerializeComponent = (props: ObjectMap = {}) => {
  return (
    <ClassComponent
      Component={FunctionComponent}
      callback={() => null}
      consumer={<ThemeContext.Consumer>{() => <div />}</ThemeContext.Consumer>}
      decorator={<WithThemeClassComponent />}
      element={<FunctionComponent {...props} />}
      forwardRef={<ForwardRefComponent />}
      fragment={<></>}
      memo={<MemoComponent />}
      portal={portal}
      provider={<ThemeContext.Provider value="" />}
      renderProp={() => <FunctionComponent {...props} renderProp={subProps => <div {...subProps} />} />}
      styled={StyledDiv}
      styledFunctionComponent={<StyledFunctionComponent />}
      styledStyledComponent={<StyledStyledComponent />}
    >
      <ForwardRefComponent />
      <FunctionComponent {...props} />
      <MemoComponent />
      <StyledFunctionComponent />
      <StyledList>
        {[1, 2, 3, 4].map(id => (
          <StyledStyledComponent key={id} />
        ))}
      </StyledList>
    </ClassComponent>
  );
};

SerializeComponent.displayName = "SerializeComponent";

export const StyledFunctionComponent = styled(FunctionComponent)``;

export const StyledStyledComponent = styled(StyledDiv)``;
