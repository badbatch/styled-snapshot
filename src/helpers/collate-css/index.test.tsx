import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import collateCSS from ".";
import { StyledItem } from "../../__test__/styled";

describe("collateCSS", () => {
  describe.only("when theme is passed in as true", () => {
    it("should return the correct structure", () => {
      const componentTree = shallow(<StyledItem />);
      const serializedTree = toJson(componentTree);
      // @ts-ignore
      const { formatted } = collateCSS(serializedTree, { theme: true });
      expect(formatted).toMatchSnapshot();
    });
  });

  describe("when theme is passed in as false", () => {
    it("should return the correct structure", () => {
      const componentTree = shallow(<StyledItem />);
      const serializedTree = toJson(componentTree);
      // @ts-ignore
      const { formatted } = collateCSS(serializedTree, { theme: false });
      expect(formatted).toMatchSnapshot();
    });
  });
});
