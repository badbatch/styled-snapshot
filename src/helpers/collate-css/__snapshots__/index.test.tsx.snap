// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collateCSS when theme is passed in as false should return the correct structure 1`] = `
"
{
  margin-left: 12px;

  position: absolute;
  padding: null;
  &:last-child {
    margin-left: 0;
  }
}
"
`;

exports[`collateCSS when theme is passed in as true should return the correct structure 1`] = `
"
{
  margin-left: 12px;

  position: relative;
  padding: 12px;
  &:last-child {
    margin-left: 0;
  }
}
"
`;
